Add GHC 9 support

AriFordsham opened the pull request:
I've modified `ghc-source-gen` so it builds and passes tests on GHC 9.0.1. I hope you find this useful for #77.

It's not production-ready; I just 'brute-forced' it - I definitely broke GHC 8.10, and I just stubbed out any new features etc. with whatever seemed to make sense to allow building.

I had to comment out two tests to get it to build; all other tests are passing.

Main issues:

- `HsTyVarBndr` now has a `Specificity` flag
- linear arrows in `infixCon``
- Various pass-dependent things have been lifted to type-level and now have less fields. Especially troublesome are the `ConPat` constructor(s)

AriFordsham commented:
@googlebot I signed it!

AriFordsham commented:
Please note I added a GHC 9 CI job and deleted the others.

blackgnezdo commented:
> Please note I added a GHC 9 CI job and deleted the others.

Is supporting at least ghc 8.10 too much trouble? I'm worried about single-target releases.

blackgnezdo commented:
> > Please note I added a GHC 9 CI job and deleted the others.
> 
> Is supporting at least ghc 8.10 too much trouble? I'm worried about single-target releases.

I think I understand now, this PR shows what it takes to get ghc 9 compatibility. Somebody needs to finish the change so that multiple versions can coexist.

This makes sense, thanks for the PR!

AriFordsham commented:
> > > Please note I added a GHC 9 CI job and deleted the others.
> > 
> > 
> > Is supporting at least ghc 8.10 too much trouble? I'm worried about single-target releases.
> 
> I think I understand now, this PR shows what it takes to get ghc 9 compatibility. Somebody needs to finish the change so that multiple versions can coexist.
> 
> This makes sense, thanks for the PR!

I don't even know if new functionality is correct. I just know it builds.

AriFordsham commented:
I think GHC 9 support #77 can be considered working experimentally at this stage: There's no new GHC 9 features, and I can't garantee I haven't broken anything making the neccesary GHC 9 changes. GHC 8 should be working unchanged.

blackgnezdo reviewed (commented):
Looks like we had a pre-existing lie:
https://www.stackage.org/lts-15.13 is for ghc-8.8.3 and so when I run your change in with ghc 8.10 I hit:
```
ghc-source-gen      >         
ghc-source-gen      > /home/greg/s/ghc-source-gen/src/GHC/SourceGen/Decl.hs:174:25: error:
ghc-source-gen      >     Data constructor not in scope: NoLayoutInfo :: NoExtField
ghc-source-gen      >     |   
ghc-source-gen      > 174 |             , tcdCExt = NoLayoutInfo 
ghc-source-gen      >     |                         ^^^^^^^^^^^^
ghc-source-gen      >         
```

I'll update the CI to be less broken.

blackgnezdo commented:
Could you rebase on top of #85 and make the tests pass?

AriFordsham commented:
I'll look at it sometime. I also remember that I left two cases commented out - I should try to fix those, or at least document if I've broken the API.

AriFordsham commented:
> All (the pull request submitter and all commit authors) CLAs are signed, **but** one or more commits were authored or co-authored by someone other than the pull request submitter.
> 
> We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that by leaving a comment that contains only `@googlebot I consent.` in this pull request.
> 
> _Note to project maintainer:_ There may be cases where the author cannot leave a comment, or the comment is not properly detected as consent. In those cases, you can manually confirm consent of the commit author(s), and set the `cla` label to `yes` (if enabled on your project).
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fgoogle%2Fghc-source-gen%2Fpull%2F84) for more info**.

@blackgnezdo

AriFordsham commented:
@blackgnezdo 8.10 CI fixed. Waiting on CLA (yours?)

AriFordsham commented on src/GHC/SourceGen/Pat/Internal.hs:
```diff
@@ -24,9 +31,15 @@ needsPar (NPat _ l _ _) = overLitNeedsParen $ unLoc l
 needsPar (NPat l _ _ _) = overLitNeedsParen $ unLoc l
 #endif
+#if MIN_VERSION_ghc(9,0,1)
+-- Think this is correct; can't be sure - AF
```
Please confirm this is correct.

AriFordsham commented on src/GHC/SourceGen/Decl.hs:
```diff
@@ -308,8 +326,16 @@ data' = newOrDataType DataType
+        name
+            $ PrefixCon
+#if MIN_VERSION_ghc(9,0,1)
+                $ map (HsScaled (HsUnrestrictedArrow NormalSyntax) . renderField)
```
`HsConDetails` of `HsConDeclDetails` now yas `HsScaled` wrapped around its first field. I've stubbed this with `(HsScaled (HsUnrestrictedArrow NormalSyntax)`

AriFordsham commented on src/GHC/SourceGen/Decl.hs:
```diff
@@ -319,7 +345,19 @@ prefixCon name fields = renderCon98Decl name
+        )
+        (
+#if MIN_VERSION_ghc(9,0,1)
+            HsScaled (HsUnrestrictedArrow NormalSyntax) $ 
```
As https://github.com/google/ghc-source-gen/pull/84/files#diff-263c7e2dfeb2bc2ce2c281b616df4b4b6b2332a1258f6cf5a05e3822bd61db05R334

AriFordsham commented on src/GHC/SourceGen/Overloaded.hs:
```diff
@@ -241,8 +258,16 @@ instance Var HsType' where
-instance BVar HsTyVarBndr' where
+#if MIN_VERSION_ghc(9,0,1)
+instance BVar HsTyVarBndrSpec' where
+    bvar = noExt UserTyVar SpecifiedSpec . typeRdrName . UnqualStr
```
No idea whether this is correct

AriFordsham commented on src/GHC/SourceGen/Pretty.hs:
```diff
@@ -4,22 +4,38 @@
+            h
+            neverQualify
+#if MIN_VERSION_ghc(9,0,1)
+                AllTheWay
```
Don't know whether this is correct

AriFordsham commented on src/GHC/SourceGen/Type.hs:
```diff
@@ -56,7 +61,13 @@ tuplePromotedTy = withPlaceHolders (noExt HsExplicitTupleTy) . map builtLoc
+a --> b =
+    noExt HsFunTy
+#if MIN_VERSION_ghc(9,0,1)
+        (HsUnrestrictedArrow NormalSyntax)
```
Again, stubbed out linear arrow type

AriFordsham commented on src/GHC/SourceGen/Decl.hs:
```diff
@@ -148,14 +164,16 @@ funDep = ClassFunDep
-#if MIN_VERSION_ghc(8,10,0)
-            , tcdCExt = NoExtField
+#if MIN_VERSION_ghc(9,0,1)
+            , tcdCExt = NoLayoutInfo
```
Don't know if this is right

AriFordsham commented on src/GHC/SourceGen/Module.hs:
```diff
@@ -49,6 +56,9 @@ module'
     -> HsModule'
 module' name exports imports decls = HsModule
     { hsmodName = fmap (builtLoc . unModuleNameStr) name
+#if MIN_VERSION_ghc(9,0,1)
```
ditto https://github.com/google/ghc-source-gen/pull/84/files#diff-263c7e2dfeb2bc2ce2c281b616df4b4b6b2332a1258f6cf5a05e3822bd61db05R174

AriFordsham reviewed (commented):
I've marked all meaningful changes.

blackgnezdo commented:
> All (the pull request submitter and all commit authors) CLAs are signed, **but** one or more commits were authored or co-authored by someone other than the pull request submitter.
> 
> We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that by leaving a comment that contains only `@googlebot I consent.` in this pull request.
> 
> _Note to project maintainer:_ There may be cases where the author cannot leave a comment, or the comment is not properly detected as consent. In those cases, you can manually confirm consent of the commit author(s), and set the `cla` label to `yes` (if enabled on your project).
> 
>  **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fgoogle%2Fghc-source-gen%2Fpull%2F84) for more info**.

Could you rebase on top of HEAD and force-push this to clear the complaint?

blackgnezdo commented:
@judah it would be great if you could look at the code as @AriFordsham has some semantic questions that will take me a while to get up to speed with.

AriFordsham commented:
I think #86 is unequivocally better than this.
